### PR TITLE
Implemented CHIP Certificate (chip-cert) Tool.

### PR DESCRIPTION
**Pull request by emargolis**
chip-cert tool provides command line interface (CLI) utility used
for generating and manipulating CHIP certificates and CHIP
private keys material.

chip-tool supports the following commands:
  - generate CHIP certificate
  - convert CHIP certificate format
  - convert CHIP private key format
  - validate CHIP certificate chain
  - resign CHIP certificate
  - print CHIP certificate

Ticket: #6035

### CODE REVIEW

File: src/tools/chip-cert/CertUtils.cpp [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+    if (ASN1_INTEGER_set_uint64(snInt, rnd) == 0)
+    {
+        ReportOpenSSLErrorAndExit("ASN1_INTEGER_set_uint64", res = false);
+    }
+
+exit:
+    return res;
+}
+
+static bool SetCertTimeField(ASN1_TIME * s, const struct tm & value)
```
Comment (mspang) [inline src/tools/chip-cert/CertUtils.cpp:201]: Can we choose a more descriptive name than `s` here?

File: src/tools/chip-cert/CertUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,826 @@
...
+    }
+
+    return res;
+}
+
+static bool AddAuthorityKeyId(X509 * cert, X509 * caCert)
+{
+    bool res = true;
+    AUTHORITY_KEYID * akid;
+    int isCritical, index = 0;
```
Comment (mspang) [inline src/tools/chip-cert/CertUtils.cpp:342]: nit: Please prefer one declaration per line.

File: src/tools/chip-cert/CertUtils.cpp [hunk 3]
```diff
@@ -0,0 +1,826 @@
...
+        {
+            ExitNow(res = false);
+        }
+        certFmt = kCertFormat_Chip_Raw;
+    }
+
+    if (certFmt == kCertFormat_Chip_Raw)
+    {
+        uint32_t convertedCertLen  = certLen * kMaxChipCertInflationFactor;
+        uint8_t * convertedCertBuf = reinterpret_cast<uint8_t *>(malloc(convertedCertLen));
```
Comment (mspang) [inline src/tools/chip-cert/CertUtils.cpp:483]: Please use `std::vector<uint8_t>` or `std::unique_ptr<uint8_t[]>` here and remove all the manual memory management with malloc/free.

File: src/tools/chip-cert/CertUtils.cpp [hunk 4]
```diff
@@ -0,0 +1,826 @@
...
+    cert = d2i_X509(nullptr, &p, certLen);
+    if (cert == nullptr)
+    {
+        ReportOpenSSLErrorAndExit("d2i_X509", res = false);
+    }
+
+exit:
+    if (cert != nullptr && !res)
+    {
+        X509_free(cert);
```
Comment (mspang) [inline src/tools/chip-cert/CertUtils.cpp:510]: Can you use `std::unique_ptr<X509, X509FreeDeleter>` where `X509FreeDeleter` is a custom deleter that calls `X509_free(ptr)` ?

File: src/tools/chip-cert/CertUtils.cpp [hunk 5]
```diff
@@ -0,0 +1,826 @@
...
+    }
+
+exit:
+    if (newCert != nullptr && !res)
+    {
+        X509_free(newCert);
+    }
+    if (!keySupplied && newKey != nullptr && !res)
+    {
+        EVP_PKEY_free(newKey);
```
Comment (mspang) [inline src/tools/chip-cert/CertUtils.cpp:787]: Can you use std::unique_ptr with a custom deleter here?

File: src/tools/chip-cert/chip-cert.h [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    kKeyFormat_Unknown = 0,
+    kKeyFormat_X509_DER,
+    kKeyFormat_X509_PEM,
+    kKeyFormat_X509_DER_PKCS8,
+    kKeyFormat_X509_PEM_PKCS8,
+    kKeyFormat_Chip_Raw,
+    kKeyFormat_Chip_Base64
+};
+
+class ToolChipDN : public chip::Credentials::ChipDN
```
Comment (mspang) [inline src/tools/chip-cert/chip-cert.h:83]: Why introduce a new type here?

**Comment by mspang** [comment]
This should be converted from C style memory & resource management to C++ RAII style.

**Comment by pan-apple** [comment]
Evgeni and I discussed over slack yesterday. In long term we can update the tool to use CHIP's certificate generation APIs (`src/credentials/GenerateChipX509Cert.cpp`). The current set of CHIP APIs are not capable enough to support all the tool's requirements.

The plan is to file issues to add more capabilities to the APIs, and then update the tool to use them.

Architecturally, the current PR looks good to me. It's a CLI tool that can be used for various offline certificate generation use cases.

**Review by pan-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by andy31415** [comment]
@emargolis - please also merge with master to remove the cirque errors (cirque was disabled in master).

File: src/tools/chip-cert/BUILD.gn [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import("//build_overrides/chip.gni")
+
+import("${chip_root}/build/chip/tools.gni")
+
+import("//build/config/linux/pkg_config.gni")
+
+pkg_config("openssl_config") {
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/tools/chip-cert/BUILD.gn:21]: Marked as TODO for the future update

File: src/tools/chip-cert/CertUtils.cpp [hunk 8]
```diff
@@ -0,0 +1,826 @@
...
+
+#include "chip-cert.h"
+
+using namespace chip;
+using namespace chip::Credentials;
+using namespace chip::ASN1;
+
+enum
+{
+    kMaxChipCertInflationFactor = 5 // Maximum ratio of the size of buffer needed to hold an X.509 certificate
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/tools/chip-cert/CertUtils.cpp:42]: Updated Max certificate enums with more realistic numbers.

File: src/tools/chip-cert/CertUtils.cpp [hunk 9]
```diff
@@ -0,0 +1,826 @@
...
+
+        fprintf(file, "%s=%s", chip::ASN1::GetOIDName(rdn[i].mAttrOID), valueStr);
+        if (certDesc != nullptr)
+        {
+            fprintf(file, " (%s)", certDesc);
+        }
+    }
+}
+
+static bool SetCertSerialNumber(X509 * cert)
```
[the remark this answers is not in the document]
Comment (emargolis) [reply src/tools/chip-cert/CertUtils.cpp:176]: Updated everywhere in this PR.

File: src/tools/chip-cert/CertUtils.cpp (hunk 1 above)
Comment (emargolis) [reply src/tools/chip-cert/CertUtils.cpp:201]: done

File: src/tools/chip-cert/CertUtils.cpp (hunk 2 above)
Comment (emargolis) [reply src/tools/chip-cert/CertUtils.cpp:342]: fixed

File: src/tools/chip-cert/CertUtils.cpp (hunk 3 above)
Comment (emargolis) [reply src/tools/chip-cert/CertUtils.cpp:483]: Updated everywhere in this PR.

File: src/tools/chip-cert/CertUtils.cpp (hunk 4 above)
Comment (emargolis) [reply src/tools/chip-cert/CertUtils.cpp:510]: done

File: src/tools/chip-cert/CertUtils.cpp (hunk 5 above)
Comment (emargolis) [reply src/tools/chip-cert/CertUtils.cpp:787]: done

**Comment by emargolis** [comment]
> src/credentials/GenerateChipX509Cert.cpp

Opened #6657 and #6656 to follow up.

**Review by jelderton — APPROVED** [review]

**Comment by woody-apple** [comment]
@saurabhst @msandstedt ?

**Review by saurabhst — APPROVED** [review]

**Pull request merged**